aten/src/ATen/native/cuda/GridSampler.cu

```diff
@@ -393,10 +396,10 @@ namespace {
 
           if (input_requires_grad) {
             // calculate and set grad_input. See Note [Passing pointer and offset to fastAtomicAdd].
-            safe_add_2d(grad_input.data, iy_nw, ix_nw, gInp_sH, gInp_sW, inp_H, inp_W, nw * gOut, NC_offset, grad_input_memory_span);
-            safe_add_2d(grad_input.data, iy_ne, ix_ne, gInp_sH, gInp_sW, inp_H, inp_W, ne * gOut, NC_offset, grad_input_memory_span);
-            safe_add_2d(grad_input.data, iy_sw, ix_sw, gInp_sH, gInp_sW, inp_H, inp_W, sw * gOut, NC_offset, grad_input_memory_span);
-            safe_add_2d(grad_input.data, iy_se, ix_se, gInp_sH, gInp_sW, inp_H, inp_W, se * gOut, NC_offset, grad_input_memory_span);
+            safe_add_2d(grad_input.data, iy_nw, ix_nw, gInp_sH, gInp_sW, inp_H, inp_W, static_cast<scalar_t>(nw * gOut), NC_offset, grad_input_memory_span);
+            safe_add_2d(grad_input.data, iy_ne, ix_ne, gInp_sH, gInp_sW, inp_H, inp_W, static_cast<scalar_t>(ne * gOut), NC_offset, grad_input_memory_span);
+            safe_add_2d(grad_input.data, iy_sw, ix_sw, gInp_sH, gInp_sW, inp_H, inp_W, static_cast<scalar_t>(sw * gOut), NC_offset, grad_input_memory_span);
+            safe_add_2d(grad_input.data, iy_se, ix_se, gInp_sH, gInp_sW, inp_H, inp_W, static_cast<scalar_t>(se * gOut), NC_offset, grad_input_memory_span);
```

Comment:
noob question: why do we `use static_cast<scalar_t>`  here but not on [L495](https://github.com/pytorch/pytorch/pull/120543/files#diff-bb063818c22c14573e6d553e9ba46402cb866917d5d93c7ac0a0831fa4e62c49R495)?

Do we need a `static_cast<scalar_t>(gOut * x_coeffs[i] * y_coeffs[j])` on line 495 as well, since that has `gOut (scalar_t), x_coeffs[i] (opmath_t),  y_coeffs[j] (opmath_t)`?

Reply:
Notice the `safe_add_2d` function signature,
```
template<typename scalar_t, typename index_t>
static __forceinline__ __device__
void safe_add_2d(scalar_t *data, int h, int w,
                 int sH, int sW, int H, int W,
                 scalar_t delta,
                 const index_t NC_offset,
                 const index_t memory_span)
```
In `safe_add_2d`, we relied on the compiler to infer what the template variable types are from the variable it received. That's why I explicitly casted the `nw * gOut` from `opmath_t` to `scalar_t`. Otherwise the compiler would be confused.

On line 495, the function is `add_value_bounded` and the function signature is
```
template<typename data_scalar_t, typename coord_scalar_t, typename index_t>
static __forceinline__ __device__
void add_value_bounded(
    data_scalar_t* data, coord_scalar_t x, coord_scalar_t y, int W, int H, int sW, int sH,
    data_scalar_t delta,
    GridSamplerPadding padding_mode,
    bool align_corners,
    const index_t NC_offset,
    const index_t memory_span)
```
We already specified the `data_scalar_t` to be `scalar_t` in the function call `add_value_bounded<scalar_t>`, therefore, the compiler would not be confused and it will implicitly cast the `gOut * x_coeffs[i] * y_coeffs[j]` from `opmath_t` to `scalar_t`. But of course, for code readability, we want everything to be explicit and we can have a `static_cast<scalar_t>` there.

Reply:
That makes a lot of sense, thank you for the detailed explanation!